File: 92. Reverse Linked List II.py

        
class Solution(object):
    def reverseBetween(self, head, left, right):
        """
        :type head: Optional[ListNode]
        :type left: int
        :type right: int
        :rtype: Optional[ListNode]
        """
        if left == right or not head: return head

        yummy = ListNode(0, head)
        node = yummy
        for i in range(0, left - 1):
            node = node.next

        tail = node.next

        for i in range(0, right - left):
            cache = tail.next
            tail.next = cache.next
            cache.next = node.next
            node.next = cache

        return yummy.next

File: 92. Reverse Linked List II.java
class ListNode{
    public int val;

    public ListNode next;

    public ListNode(int val, ListNode next){
        this.val = val;
        this.next = next;
    }
}

class Solution{
    public ListNode reverseBetween(ListNode head, int left, int right){
        if(left == right || head == null) return head;

        ListNode yummy = new ListNode(0, head);
        ListNode node = yummy;

        for(int i = 0; i < left - 1; ++i) node = node.next;

        ListNode tail = node.next;

        for(int i = 0; i < right - left; ++i){
            ListNode cache = tail.next;
            tail.next = cache.next;
            cache.next = node.next;
            node.next = cache;
        }

        return yummy.next;
    }
}